File: rjh_agent/tools/tavily_search.py
from __future__ import annotations

import json
import urllib.error
import urllib.request
from dataclasses import dataclass


@dataclass
class SearchResult:
    title: str
    url: str
    content: str
    score: float


class TavilySearch:
    def __init__(self, api_key: str):
        self.api_key = api_key
        self.base_url = "https://api.tavily.com/search"

    def search(self, query: str, max_results: int = 5) -> list[SearchResult]:
        print(f"[TavilySearch] 搜索：{query}")

        payload = {
            "query": query,
            "max_results": max_results,
            "search_depth": "basic",
            "include_answer": False,
            "include_raw_content": False,
        }

        request = urllib.request.Request(
            self.base_url,
            data=json.dumps(payload).encode("utf-8"),
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {self.api_key}",
            },
            method="POST",
        )

        try:
            with urllib.request.urlopen(request, timeout=30) as response:
                data = json.loads(response.read().decode("utf-8"))
        except urllib.error.URLError as err:
            print(f"[TavilySearch] 搜索失败：{err}")
            return []

        results = [
            SearchResult(
                title=item.get("title", ""),
                url=item.get("url", ""),
                content=item.get("content", "")[:800],
                score=float(item.get("score", 0)),
            )
            for item in data.get("results", [])
        ]

        print(f"[TavilySearch] 获取到 {len(results)} 条结果")
        return results

File: src/tools/tavily-search.ts
/**
 * Tavily 搜索工具
 * 专为 AI 智能体优化的搜索引擎，适合在 Agent 中使用
 * 获取 Key：https://app.tavily.com/ （免费额度每月 1000 次）
 */

export interface SearchResult {
  title: string
  url: string
  content: string
  score: number
}

export class TavilySearch {
  private apiKey: string
  private baseUrl = 'https://api.tavily.com/search'

  constructor(apiKey: string) {
    this.apiKey = apiKey
  }

  /**
   * 执行搜索
   * @param query 搜索关键词
   * @param maxResults 最大返回数量，默认 5
   */
  async search(query: string, maxResults = 5): Promise<SearchResult[]> {
    console.log(`[TavilySearch] 搜索：${query}`)

    try {
      const response = await fetch(this.baseUrl, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${this.apiKey}`,
        },
        body: JSON.stringify({
          query,
          max_results: maxResults,
          search_depth: 'basic',
          include_answer: false,
          include_raw_content: false,
        }),
      })

      if (!response.ok) {
        throw new Error(`Tavily API 错误：${response.status} ${response.statusText}`)
      }

      const data = (await response.json()) as {
        results: Array<{
          title: string
          url: string
          content: string
          score: number
        }>
      }

      console.log(`[TavilySearch] 获取到 ${data.results.length} 条结果`)

      return data.results.map(r => ({
        title: r.title,
        url: r.url,
        content: r.content.slice(0, 800), // 截取前 800 字符，控制 Token 用量
        score: r.score,
      }))
    } catch (err) {
      console.error('[TavilySearch] 搜索失败：', err)
      return []
    }
  }
}
